checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -141,6 +327,123 @@ private void visitAccess(ExpressionTree indexTree, ExpressionTree arrTree) {
+     * if the reassignment might invalidate any dependent types in scope.
+     */
+    @Override
+    protected void commonAssignmentCheck(
+            Tree varTree, ExpressionTree valueExp, @CompilerMessageKey String errorKey) {
+        super.commonAssignmentCheck(varTree, valueExp, errorKey);
+        if (varTree.getKind() == Kind.VARIABLE) {
+            // Not a reassignment, so nothing to check
+            return;
+        }
+        Element elt = InternalUtils.symbol(varTree);
+
+        if (elt != null
+                && (elt.getKind() == ElementKind.FIELD
+                        || elt.getKind() == ElementKind.PARAMETER
+                        || elt.getKind() == ElementKind.LOCAL_VARIABLE)
+                && !ElementUtils.isEffectivelyFinal(elt)) {
+            TypeKind typeKind = InternalUtils.typeOf(varTree).getKind();
+            Receiver field = FlowExpressions.internalReprOf(atypeFactory, (ExpressionTree) varTree);
+            ReassignmentKind kind;
+            if (elt.getKind() == ElementKind.FIELD) {
+                if (typeKind == TypeKind.ARRAY) {
+                    kind = ReassignmentKind.ARRAY_FIELD_REASSIGNMENT;
+                } else if (!typeKind.isPrimitive()) {
+                    kind = UpperBoundUtil.ReassignmentKind.NON_ARRAY_FIELD_REASSIGNMENT;
+                } else {
+                    return;
+                }
+            } else if (typeKind == TypeKind.ARRAY) {
+                kind = UpperBoundUtil.ReassignmentKind.LOCAL_VAR_REASSIGNMENT;
+            } else {
+                return;
+            }
+            checkAnnotationsInClass(elt, field, varTree, getCurrentPath(), kind);
+        }
+    }
+
+    /**
+     * Checks whether any annotations in scope are dependent and need to be invalidated by the given
+     * side effect.
+     *
+     * @param elt An element corresponding to the side effect that caused the invalidation.
+     * @param possiblyInvalidatedRec A receiver representing the possibly invalidated expression.
+     * @param tree The tree on which the side effect occurs. Used in error reporting.
+     * @param path The path on which the side effect occurs. Used for parsing.
+     * @param reassignmentKind The type of side effect. See {@link UpperBoundUtil.ReassignmentKind}.
+     */
+    private void checkAnnotationsInClass(
+            Element elt,
+            Receiver possiblyInvalidatedRec,
+            Tree tree,
+            TreePath path,
+            UpperBoundUtil.ReassignmentKind reassignmentKind) {
+        List<? extends Element> enclosedElts =
+                enclosingElementsCache.containsKey(elt) ? enclosingElementsCache.get(elt) : null;
+        if (enclosedElts == null) {
+            enclosedElts = ElementUtils.enclosingClass(elt).getEnclosedElements();
+            enclosingElementsCache.put(elt, enclosedElts);
+        }
+        List<AnnotatedTypeMirror> enclosedTypes = findEnclosedTypes(enclosedElts);
```

Comment:
Here's another comment from the last review that was lost:
https://github.com/typetools/checker-framework/pull/1346#discussion_r143769405